flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -311,71 +310,45 @@ public void testCancellationFailsWithBlockingLock() throws Exception {
-		OperatorID operatorID3 = new OperatorID();
-		when(streamOperator1.getOperatorID()).thenReturn(operatorID1);
-		when(streamOperator2.getOperatorID()).thenReturn(operatorID2);
-		when(streamOperator3.getOperatorID()).thenReturn(operatorID3);
-
-		// set up the task
-
-		StreamOperator<?>[] streamOperators = {streamOperator1, streamOperator2, streamOperator3};
+		LinkedBlockingQueue<Event> eventQueue = new LinkedBlockingQueue<>();
 
-		OperatorChain<Void, AbstractStreamOperator<Void>> operatorChain = mock(OperatorChain.class);
-		when(operatorChain.getAllOperators()).thenReturn(streamOperators);
+		try (MockEnvironment mockEnvironment = new MockEnvironmentBuilder().build()) {
+			RunningTask<EmptyStreamTask> task = runTask(() -> new EmptyStreamTask(
+				mockEnvironment,
+				eventQueue,
+				operatorChain(
+					streamOperatorWithSnapshot(operatorSnapshotResult1),
+					streamOperatorWithSnapshot(operatorSnapshotResult2),
+					streamOperatorWithSnapshotException(testException))));
+			EmptyStreamTask streamTask = task.streamTask;
 
-		Whitebox.setInternalState(streamTask, "isRunning", true);
-		Whitebox.setInternalState(streamTask, "lock", new Object());
-		Whitebox.setInternalState(streamTask, "operatorChain", operatorChain);
-		Whitebox.setInternalState(streamTask, "cancelables", new CloseableRegistry());
-		Whitebox.setInternalState(streamTask, "configuration", new StreamConfig(new Configuration()));
-		Whitebox.setInternalState(streamTask, "checkpointStorage", new MemoryBackendCheckpointStorage(new JobID(), null, null, Integer.MAX_VALUE));
+			assertThat(eventQueue.take(), is(Event.TASK_IS_RUNNING));
+			assertTrue(eventQueue.isEmpty());
 
-		CheckpointExceptionHandlerFactory checkpointExceptionHandlerFactory = new CheckpointExceptionHandlerFactory();
-		CheckpointExceptionHandler checkpointExceptionHandler =
-			checkpointExceptionHandlerFactory.createCheckpointExceptionHandler(true, mockEnvironment);
-		Whitebox.setInternalState(streamTask, "synchronousCheckpointExceptionHandler", checkpointExceptionHandler);
+			try {
+				streamTask.triggerCheckpoint(
+					new CheckpointMetaData(42L, 1L),
+					CheckpointOptions.forCheckpointWithDefaultLocation(),
+					false);
+
+				fail("Expected test exception here.");
+			} catch (Exception e) {
+				assertThat(e.getCause(), is(testException));
+			}
 
-		StreamTask.AsyncCheckpointExceptionHandler asyncCheckpointExceptionHandler =
-			new StreamTask.AsyncCheckpointExceptionHandler(streamTask);
-		Whitebox.setInternalState(streamTask, "asynchronousCheckpointExceptionHandler", asyncCheckpointExceptionHandler);
+			verify(operatorSnapshotResult1).cancel();
+			verify(operatorSnapshotResult2).cancel();
 
-		try {
-			streamTask.triggerCheckpoint(checkpointMetaData, CheckpointOptions.forCheckpointWithDefaultLocation(), false);
-			fail("Expected test exception here.");
-		} catch (Exception e) {
-			assertEquals(testException, e.getCause());
+			task.streamTask.setSourceFinished();
```

Comment:
Why are you calling `setSourceFinished` only 2 times? In other places task is failing exceptionally/being cancelled? 

Also I think better name would be "input" not "source".

If so maybe rename the method to `finishInput()`? `finishTaskGracefully()`? Or at least add a java doc explaining what this method is doing.

Reply:
This is just preserving old tests semantics as much as possible, although, it may not really matter for the provided tests.

+1 for renaming.